src/CallManager.cpp

```diff
@@ -0,0 +1,315 @@
+    hangUp();
+    return;
+  }
+  session_.acceptICECandidates(remoteICECandidates_);
+  remoteICECandidates_.clear();
+}
+
+void
+CallManager::handleEvent(const RoomEvent<CallCandidates> &callCandidatesEvent)
+{
+  nhlog::ui()->debug("CallManager::incoming CallCandidates from {} with id {}", callCandidatesEvent.sender, callCandidatesEvent.content.call_id);
+  if (callid_ == callCandidatesEvent.content.call_id) {
+    if (onActiveCall())
+      session_.acceptICECandidates(callCandidatesEvent.content.candidates);
+    else {
+      // CallInvite has been received and we're awaiting localUser to accept or reject the call
+      for (const auto &c : callCandidatesEvent.content.candidates)
+        remoteICECandidates_.push_back(c);
+    }
+  }
+}
+
+void
+CallManager::handleEvent(const RoomEvent<CallAnswer> &callAnswerEvent)
+{
+  nhlog::ui()->debug("CallManager::incoming CallAnswer from {} with id {}", callAnswerEvent.sender, callAnswerEvent.content.call_id);
+  if (onActiveCall() && callid_ == callAnswerEvent.content.call_id) {
+    stopRingtone();
+    if (!session_.acceptAnswer(callAnswerEvent.content.sdp)) {
+      emit ChatPage::instance()->showNotification("Problem setting up call");
+      hangUp();
+    }
+  }
+}
+
+void
+CallManager::handleEvent(const RoomEvent<CallHangUp> &callHangUpEvent)
+{
+  nhlog::ui()->debug("CallManager::incoming CallHangUp from {} with id {}", callHangUpEvent.sender, callHangUpEvent.content.call_id);
+  if (onActiveCall() && callid_ == callHangUpEvent.content.call_id)
+    endCall();
+}
+
+void
+CallManager::generateCallID()
+{
+  using namespace std::chrono;
+  uint64_t ms = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+  callid_ = "c" + std::to_string(ms);
+}
+
+void
+CallManager::endCall()
+{
+  stopRingtone();
+  session_.end();
+  roomid_.clear();
+  callid_.clear();
+  remoteICECandidates_.clear();
+}
+
+void
+CallManager::retrieveTurnServer()
+{
+  http::client()->get_turn_server(
+    [this](const mtx::responses::TurnServer &res, mtx::http::RequestErr err) {
+        if (err) {
+            turnServerTimer_.setInterval(5000);
+            return;
+        }
+        emit turnServerRetrieved(res);
+  });
+}
+
+void
+CallManager::setTurnServers()
+{
+  // gstreamer expects (percent-encoded): turn(s)://username:password@host:port?transport=udp(tcp)
+  std::vector<std::string> uris;
+  for (const auto &uri : turnServer_.uris) {
+    if (auto c = uri.find(':'); c == std::string::npos) {
+      nhlog::ui()->error("Invalid TURN server uri: {}", uri);
+      continue;
+    }
+    else {
+      std::string scheme = std::string(uri, 0, c);
+      if (scheme != "turn" && scheme != "turns") {
+        nhlog::ui()->error("Invalid TURN server uri: {}", uri);
+        continue;
+      }
+      std::string res = scheme + "://" + turnServer_.username + ":" + turnServer_.password
+        + "@" + std::string(uri, ++c);
```

Comment:
You need to percent encode the fragments, not the entire URL. Also, until https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/issues/831 is fixed, this will still not work sadly, but there is an open MR for it since a few days.

Reply:
```
      QString encodedUri = QString::fromStdString(scheme) + "://" +
                           QUrl::toPercentEncoding(QString::fromStdString(turnServer_.username)) +
                           ":" +
                           QUrl::toPercentEncoding(QString::fromStdString(turnServer_.password)) +
                           "@" + QString::fromStdString(std::string(uri, ++c));
      uris.push_back(encodedUri.toStdString());
```

Something like this + the patch for gstreamer seems to at least give no turn server errors anymore (although it still fails to actually connect).